### PR TITLE
[COLLECTIONS 708] Add hashCode method to CollectionUtils

### PR DESCRIPTION
**Pull request by dota17**
For [COLLECTIONS 708](https://issues.apache.org/jira/projects/COLLECTIONS/issues/COLLECTIONS-708?filter=allopenissues)

### CODE REVIEW

File: src/main/java/org/apache/commons/collections4/CollectionUtils.java [hunk 1]
```diff
@@ -608,6 +608,33 @@ public static boolean isEqualCollection(final Collection<?> a, final Collection<
...
+     * </p>
+     *
+     * @param <E>  the element type
+     * @param collection  the input collection
+     * @param equator  the Equator used for generate hashCode
+     * @return the hashCode of Collection through the hash method of Equator
+     */
+    public static <E> int hashCode( final Collection<? extends E> collection,
+                                    final Equator<? super E> equator ) {
+        if ( null == equator ) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/CollectionUtils.java]: No spaces inside parens please: `if ( foo )` -> `if (foo)`.

File: src/main/java/org/apache/commons/collections4/CollectionUtils.java [hunk 2]
```diff
@@ -608,6 +608,33 @@ public static boolean isEqualCollection(final Collection<?> a, final Collection<
...
+            throw new NullPointerException( "Equator must not be null." );
+        }
+        if ( null == collection ){
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/CollectionUtils.java]: `){` -> `) {`

File: src/main/java/org/apache/commons/collections4/CollectionUtils.java [hunk 3]
```diff
@@ -608,6 +608,33 @@ public static boolean isEqualCollection(final Collection<?> a, final Collection<
 
         return isEqualCollection(collect(a, transformer), collect(b, transformer));
     }
+    
+    /**
+     * Getting the hashCode of Collection through the hash method of Equator.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/CollectionUtils.java]: Please replace `Getting the hashCode` with `Returns the hash code`.

File: src/main/java/org/apache/commons/collections4/CollectionUtils.java [hunk 4]
```diff
@@ -608,6 +608,33 @@ public static boolean isEqualCollection(final Collection<?> a, final Collection<
...
+     * <p>
+     * If the input collection is null return 1.
+     * </p>
+     *
+     * @param <E>  the element type
+     * @param collection  the input collection
+     * @param equator  the Equator used for generate hashCode
+     * @return the hashCode of Collection through the hash method of Equator
+     */
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/CollectionUtils.java]: Please add a `@since` tag since this a new public method.

File: src/main/java/org/apache/commons/collections4/CollectionUtils.java [hunk 5]
```diff
@@ -608,6 +608,33 @@ public static boolean isEqualCollection(final Collection<?> a, final Collection<
...
+                                   final Equator<? super E> equator) {
+        if (null == equator) {
+            throw new NullPointerException( "Equator must not be null." );
+        }
+        if (null == collection) {
+            return 1;
+        }
+        int hashCode = 1;
+        for (final E e : collection) {
+
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/CollectionUtils.java]: Remove this extra blank line please.

**Comment by dota17** [comment]
> 我认为你缺少输入上的null集合的测试用例。请运行网站构建并检查新代码是否获得100％的代码覆盖率`mvn clean site -P jacoco`。
> ...谢谢！

Sorry, this is the first time I have submitted code for Apache project. Many details are problematic.

**Comment by dota17** [comment]
> I think you are missing a test case for a null collection on input. Please run the site build and check that the new code gets 100% code coverage `mvn clean site -P jacoco`.
> ... and thank you!

Sorry, this is the first time I have submitted code for Apache project. Many details are problematic.

File: src/test/java/org/apache/commons/collections4/CollectionUtilsTest.java [hunk 6]
```diff
@@ -620,6 +620,52 @@ public int hash(final Integer o) {
...
+            public boolean equate(final Integer o1, final Integer o2) {
+                if (o1 % 2 == 0 ^ o2 %2 == 0) {
+                    return false;
+                }
+                return true;
+            }
+
+            @Override
+            public int hash(final Integer o) {
+                return o == null ? 0 : o.hashCode();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/CollectionUtilsTest.java]: Replace this with a call to `java.util.Objects.hashCode(Object)`.

File: src/test/java/org/apache/commons/collections4/CollectionUtilsTest.java [hunk 7]
```diff
@@ -620,6 +620,52 @@ public int hash(final Integer o) {
...
+            public boolean equate(final Integer o1, final Integer o2) {
+                if (o1 % 2 == 0 ^ o2 % 2 == 0 ) {
+                    return false;
+                }
+                return true;
+            }
+
+            @Override
+            public int hash(final Integer o) {
+                return o == null ? 0 : o.hashCode();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/collections4/CollectionUtilsTest.java]: java.util.Objects.hashCode(Object)
